crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+        // which creates a nightmare with lifetimes unless you own the archive, hence the clone.
+        //
+        // The zip documentation states that:
+        //
+        //     At the moment, this type is cheap to clone if this is the case for the reader it uses.
+        //     However, this is not guaranteed by this crate and it may change in the future.
+        //
+        // So this should be fairly cheap *for now*.
+        self.inner.clone()
+    }
+
+    /// Create a new empty string, read the contents of `zipfile` into it, and return the string
+    fn read_zipfile(mut zipfile: ZipFile) -> Result<String> {
+        let mut buffer = String::with_capacity(zipfile.size() as usize);
+        zipfile.read_to_string(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+impl FileSystem for VendoredFileSystem {
+    fn exists(&self, path: &FileSystemPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        self.vendored_archive().lookup_path(&normalized).is_ok()
+            || (!normalized.has_trailing_slash()
+                && self
+                    .vendored_archive()
+                    .lookup_path(&normalized.with_trailing_slash())
+                    .is_ok())
+    }
+
+    fn is_directory(&self, path: &FileSystemPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        if let Ok(zipfile) = self.vendored_archive().lookup_path(&normalized) {
+            return zipfile.is_dir();
+        }
+        if normalized.has_trailing_slash() {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalized.with_trailing_slash())
+            .is_ok_and(|zipfile| zipfile.is_dir())
+    }
+
+    fn is_file(&self, path: &FileSystemPath) -> bool {
+        if path.as_str().ends_with('/') {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalize_vendored_path(path))
+            .is_ok_and(|zipfile| zipfile.is_file())
+    }
+
+    fn metadata(&self, path: &FileSystemPath) -> Result<Metadata> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+
+        let is_dir = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => zipfile.is_dir(),
+            Err(err) => {
+                if normalized.has_trailing_slash() {
+                    return Err(err);
+                }
+                let mut archive = self.vendored_archive();
+                let lookup_result = archive.lookup_path(&normalized.with_trailing_slash());
+                if let Ok(zipfile) = lookup_result {
+                    zipfile.is_dir()
+                } else {
+                    return Err(err);
+                }
+            }
+        };
+
+        let file_type = if is_dir {
+            FileType::Directory
+        } else {
+            FileType::File
+        };
+
+        Ok(Metadata {
+            revision: FileRevision::zero(),
```

Comment:
We should use a revision other than zero or we might run into problems when the vendored file system content changes. We could use ruff's version, the last modified time of the files in the zip file, or the time when the `VendoredFileSystem` was created.

Reply:
I thought one of the guarantees of the vendored file system was that the files could never change within a single invocation of Ruff. Or is the concern here that this would lead to incorrect data being loaded from the persistent on-disk cache on a subsequent run of Ruff?

Reply:
It should probably be fine without it, because we should bust the entire cache if the ruff version changes. But having it in the revision gives us an extra layer of protection, because salsa would then invalidate the queries for us, in case we forget.